### PR TITLE
feat: Support gRPC Protocol for Beagle Web through a proxy server

### PR DESCRIPTION
**Pull request by arthurbleilzup**

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">The title of PR not follow the rules of: https://www.conventionalcommits.org/en/v1.0.0/</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please provide a summary in the Pull Request description</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: web/beagle-grpc/src/lib/utils/request.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  if (requestInfo) {
+    if (typeof requestInfo === 'string') {
+      return { url: requestInfo }
+    } else if (typeof requestInfo === 'object') {
+      const { url, method, headers, body } = requestInfo
+      return { url, method, headers, body }
+    }
+  }
+  return null
+}
```
Comment (Tiagoperes) [inline web/beagle-grpc/src/lib/utils/request.ts:17]: Too many nested ifs. Suggestion:

```typescript
if (typeof requestInfo === 'string') return { url: requestInfo }
if (requestInfo && typeof requestInfo === 'object') {
  const { url, method, headers, body } = requestInfo
  return { url, method, headers, body }
}
return null
```

File: web/beagle-grpc/__tests__/lib/element.spec.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      expect(beagleUIElement._beagleComponent_).toBe(mockId)
+    })
+
+    test('it should return a BeagleUIElement with context', () => {
+      expect(beagleUIElement.context).toBeDefined()
+      expect(beagleUIElement.context?.id).toBe('mock-context-id')
+      expect(beagleUIElement.context?.value).toEqual({ value: 'mock-context-value' })
+    })
+
+    test('it should return a BeagleUIElement with style', () => {
```
Comment (Tiagoperes) [inline web/beagle-grpc/__tests__/lib/element.spec.ts:49]: Have you switched the implementation of this test with the next one?

**Review by Tiagoperes — CHANGES_REQUESTED** [review]
Wow, so many tests! Good work!

As a general commentary, sometimes you throw errors as strings, without creating an object that extends the Error class. This is unusual. You also sometimes create the error class. I didn't notice any pattern for using or not `Error(message)`. Can you please explain?

Now we need a good documentation for this, because GRPC can't work natively for WEB, I imagine our "installation guide" will be more complicated than the others.

Also, we need a github action to run all these tests...

File: web/beagle-grpc/__tests__/lib/element.spec.ts (hunk 2 above)
Comment (arthurbleilzup) [reply web/beagle-grpc/__tests__/lib/element.spec.ts:49]: yes...

**Comment by arthurbleilzup** [comment]
> Wow, so many tests! Good work!
> 
> As a general commentary, sometimes you throw errors as strings, without creating an object that extends the Error class. This is unusual. You also sometimes create the error class. I didn't notice any pattern for using or not `Error(message)`. Can you please explain?
> 
> Now we need a good documentation for this, because GRPC can't work natively for WEB, I imagine our "installation guide" will be more complicated than the others.
> 
> Also, we need a github action to run all these tests...

Because when I was throwing those errors I just logged them and the application was exited, but right now with your point of view, it make sense to be an Error object, so I fixed it

**Review by Tiagoperes — DISMISSED** [review]

**Review by Tiagoperes — APPROVED** [review]

**Pull request merged**